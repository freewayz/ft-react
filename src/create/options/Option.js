import React from 'react'
import Answer from './Answer'
import FeedBack from './FeedBack'

class Option extends React.Component {

	state = {
		is_correct: this.props.option.is_correct
	}

	handleIsCorrect = () => {
		const checked = !this.state.is_correct
		this.setState(({is_correct}) => ({is_correct: checked}))
		this.optionChange({is_correct: checked})

	}

	optionChange = (value) => {
		// update redux store with any of children 
		// component value change (Answer,FeedBack and Checkbox)
		const {id,onUpdate} = this.props
		onUpdate({id, ...value})
	}
	render() {
		const {option} = this.props
		const {is_correct} = this.state
		const commonProps = {
			handleUpdate: this.optionChange
		}
		const panelStyle = is_correct ? 'is__correct': ''

		return (
			<div className='option__panel'>
				<div className={`option__panel__header ${panelStyle}`}>
					<span className='option__label'> A </span>
					<span> {is_correct ? 'Correct' : 'Mark as correct'}
			<input 
							type='checkbox' 
							onChange={this.handleIsCorrect}
							checked={this.state.is_correct} />
					</span>
				</div>
				<div className='option__panel__content'>
					{/* Handle compoenent differently so we can test and refactor with ease */}
				<Answer info={option.answer} handleUpdate={this.optionChange} />
				<FeedBack info={option.feedback} {...commonProps} />
			</div>
			</div>
		)
	}
}

export default Option
