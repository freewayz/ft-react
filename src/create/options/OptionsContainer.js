import React from 'react'
import {connect} from 'react-redux'
import PropTypes from 'prop-types'
import {editOption, removeOption} from './../../actions'
import Option from './Option'

class OptionsContainer extends React.Component {

	handleUpdateOptionEdit = (details) => {
		// the index of the option
		this.props.editOption({details})
	}

	render () {
		const {options, removeOption} = this.props
		return (
			<div className='question__options'>
				<h2>Answers </h2>
				<div className='options'>
					{
						options.map((option, index) => <Option
							option={option} 
							id={index}
							removeOption={removeOption}
							onUpdate={this.handleUpdateOptionEdit}
							key={index} />
						)
					}
				</div>
			</div>
		)
	}
}

OptionsContainer.propTypes = {
	options: PropTypes.array,
	editOption: PropTypes.func.isRequired,
	removeOption: PropTypes.func.isRequired
}

const mapDispatchToProps = {
	editOption,
	removeOption
}

export default connect(null, mapDispatchToProps)(OptionsContainer)
