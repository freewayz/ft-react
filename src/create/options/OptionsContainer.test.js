/* global expect describe it shallow mount  beforeEach */
import React from 'react'
import configureStore from 'redux-mock-store'
import OptionsContainer from './OptionsContainer'
import Option from './Option'

const initialState = {}

const mockStore = configureStore()
const options = [
  {
    is_correct: false,
    answer: 'Answer 1',
    feedback: 'Feedback 1'
  },
  {
    is_correct: true,
    answer: 'Answer 2',
    feedback: 'Feedback 2'
  }
]
let store
let wrapper

beforeEach(() => {
  store = mockStore(initialState)
  const context = {store}
  wrapper = shallow(<OptionsContainer options={options} />, {context})
})

describe('OptionsContainer Component TestCase', () => {
  it('renders the right snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 2 children for Option', () => {
    expect(wrapper.dive().find(Option).length).toEqual(2)
  })
})
