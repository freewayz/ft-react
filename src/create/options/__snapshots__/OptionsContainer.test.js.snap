// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsContainer Component TestCase renders the right snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(OptionsContainer)
    options={
        Array [
            Object {
              "answer": "Answer 1",
              "feedback": "Feedback 1",
              "is_correct": false,
            },
            Object {
              "answer": "Answer 2",
              "feedback": "Feedback 2",
              "is_correct": true,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "editOption": [Function],
      "options": Array [
        Object {
          "answer": "Answer 1",
          "feedback": "Feedback 1",
          "is_correct": false,
        },
        Object {
          "answer": "Answer 2",
          "feedback": "Feedback 2",
          "is_correct": true,
        },
      ],
      "removeOption": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "editOption": [Function],
        "options": Array [
          Object {
            "answer": "Answer 1",
            "feedback": "Feedback 1",
            "is_correct": false,
          },
          Object {
            "answer": "Answer 2",
            "feedback": "Feedback 2",
            "is_correct": true,
          },
        ],
        "removeOption": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
